### PR TITLE
feat: implement sampling support for Streamable HTTP transport

### LINKED ISSUE #419: Feature: Add Sampling Support
**Priority**: High  
**Breaking Change**: No

**Description**: Implement MCP sampling support to allow servers to request LLM sampling ("completions" or "generations") from language models via clients. This enables servers to implement agentic behaviors with nested LLM calls while maintaining client control over model access and permissions.

## Overview
The sampling feature allows servers to request text, audio, or image-based interactions from language models through the client, with no server API keys necessary. Clients maintain control over model selection, permissions, and user approval.

## Key Features
- **sampling/createMessage** request support
- Model preference system with capability priorities and hints
- Support for text, image, and audio content
- User approval controls and security considerations
- Rate limiting and sensitive data handling

### Core Implementation
- [ ] Add sampling capability declaration during initialization
- [ ] Implement `sampling/createMessage` request/response handling
- [ ] Add message content types (text, image, audio)
- [ ] Implement model preference system with priorities and hints
- [ ] Add proper error handling for sampling requests
- [ ] Implement security controls and user approval mechanisms

#### STDIO Transport
- [ ] Implement sampling support in STDIO client transport
- [ ] Add bidirectional message handling for sampling requests
- [ ] Test sampling flow with STDIO server examples
- [ ] Add STDIO-specific error handling for sampling

#### Streamable HTTP Transport  
- [ ] Implement sampling support in Streamable HTTP client transport
- [ ] Add HTTP-specific headers and request handling for sampling
- [ ] Test sampling flow with HTTP server examples
- [ ] Add HTTP-specific error handling and status codes for sampling

### Data Types and Validation
- [ ] Add `SamplingMessage` types with role and content fields
- [ ] Add `ModelPreferences` struct with priorities and hints
- [ ] Add `SamplingResult` struct with model and stop reason
- [ ] Implement content validation for text, image, and audio
- [ ] Add JSON schema validation for sampling requests

### Security and User Experience
- [ ] Implement user approval UI/controls for sampling requests
- [ ] Add prompt review and editing capabilities
- [ ] Add response review before delivery
- [ ] Implement rate limiting for sampling requests
- [ ] Add logging and audit trail for sampling operations

### Testing and Documentation
- [ ] Add unit tests for sampling message types
- [ ] Add integration tests for both transport types
- [ ] Add examples demonstrating sampling usage
- [ ] Update documentation with sampling capabilities
- [ ] Add security best practices documentation

## Files to Update
- `mcp/types.go` - Add sampling message types
- `client/interface.go` - Add sampling methods
- `client/stdio.go` - STDIO sampling implementation
- `client/http.go` - HTTP sampling implementation  
- `server/server.go` - Server-side sampling support
- `examples/` - Add sampling examples

## Security Considerations
1. Clients SHOULD implement user approval controls
2. Both parties SHOULD validate message content
3. Clients SHOULD respect model preference hints
4. Clients SHOULD implement rate limiting
5. Both parties MUST handle sensitive data appropriately

## Reference
Based on MCP Specification 2025-06-18: https://modelcontextprotocol.io/specification/2025-06-18/client/sampling

### PR DESCRIPTION
**Pull request by andig**
Implements sampling capability for HTTP transport, resolving issue #419.

## Todo

- [ ] decide explicit timeouts instead of chosen values

## What's New

- **Bidirectional HTTP Communication**: Servers can send sampling requests to HTTP clients via SSE
- **Client Response Handling**: Clients respond via HTTP POST with proper session correlation
- **Complete Interface Support**: Implements `BidirectionalInterface` for HTTP transport
- **Working Examples**: Both client and server examples with comprehensive documentation

## Key Features

- Uses existing MCP headers (`Mcp-Session-Id`, `Mcp-Protocol-Version`)
- JSON-RPC 2.0 compliant request/response handling
- Proper error codes for different failure scenarios
- Thread-safe with goroutine-based async processing
- Full backward compatibility

## Architecture

```
Server (HTTP) ←→ Client (HTTP + SSE)
     │              │
     └─ POST ←─── SSE Stream ───→ Sampling Request
     │              │
     └─ Response ──→ POST Response ───→ Server
```

## Testing

All tests pass ✅ including comprehensive coverage for:
- Complete sampling flow scenarios
- Error handling and edge cases  
- Interface compliance verification
- HTTP-specific error codes

## Usage

```go
// Create HTTP transport with sampling support
httpTransport, err := transport.NewStreamableHTTP("http://localhost:8080")
mcpClient := client.NewClient(httpTransport, client.WithSamplingHandler(samplingHandler))
```

Closes #419

## Summary by CodeRabbit

* **New Features**
  * Added support for bidirectional sampling requests and responses over HTTP, enabling server-initiated sampling and client handling of these requests.
  * Introduced example HTTP client and server programs demonstrating sampling capabilities, including mock LLM integration.
  * Server and client now advertise and detect sampling capability during initialization.

* **Bug Fixes**
  * Improved error handling for sampling requests and responses, including proper status codes and error messaging.
  * Enhanced context-aware timeouts and cancellation handling to improve connection stability.

* **Documentation**
  * Added comprehensive README guides for both the sampling HTTP client and server examples, including usage, features, and production considerations.

* **Tests**
  * Introduced extensive unit tests for sampling request/response flows, error scenarios, interface compliance, and queue management for both server and client components.

* **Chores**
  * Added graceful shutdown handling to the client example for better resource cleanup on termination signals.

### CODE REVIEW

**Comment by andig** [comment]
**Please be cautious:** I've Claude do the work and only took a cursory glance. I cannot fully validate the result!

**Comment by andig** [comment]
I don't think we can accept this: https://github.com/mark3labs/mcp-go/pull/515/files#diff-295aeddf58426d87ebc2fa3cf760bc3877dce12ad2e9bc3735c144e310dd98c7R589-R592

**Comment by andig** [comment]
## Review Comments Fixed ✅

I've addressed all the unresolved review comments with the following commits:

### 1. [Fix graceful shutdown handling in sampling client](https://github.com/mark3labs/mcp-go/commit/a9b20be1)
- ✅ Added signal handling for SIGINT and SIGTERM  
- ✅ Moved defer statement after error checking
- ✅ Improved shutdown error handling with proper logging

### 2. [Improve context handling in streamable HTTP transport](https://github.com/mark3labs/mcp-go/commit/b7afbb94)
- ✅ Added timeout context for SSE response processing (30s default)
- ✅ Added timeout for individual connection attempts in listenForever (10s)
- ✅ Used context-aware sleep in retry logic
- ✅ Ensured async goroutines properly respect context cancellation

### 3. [Fix server streamable HTTP issues](https://github.com/mark3labs/mcp-go/commit/a6642894)
- ✅ Improved error message for queue full condition to be more descriptive and actionable
- ✅ All session struct fields are in use (no unused fields found)
- ✅ Notification sending already uses proper non-blocking approach with error handling

### 4. [Rename struct variables for clarity](https://github.com/mark3labs/mcp-go/commit/bac5dadc)  
- ✅ Renamed `responseMessage` to `jsonMessage` for more neutral naming
- ✅ Enhanced code readability and follows consistent naming conventions

All fixes maintain backward compatibility and follow established patterns in the codebase. The changes improve error handling, context management, and code clarity as requested in the review comments.

File: server/streamable_http.go [hunk 1]
```diff
@@ -223,14 +224,32 @@ func (s *StreamableHTTPServer) handlePost(w http.ResponseWriter, r *http.Request
...
 	}
-	if err := json.Unmarshal(rawData, &baseMessage); err != nil {
+	if err := json.Unmarshal(rawData, &jsonMessage); err != nil {
 		s.writeJSONRPCError(w, nil, mcp.PARSE_ERROR, "request body is not valid json")
 		return
 	}
-	isInitializeRequest := baseMessage.Method == mcp.MethodInitialize
+
+	// Check if this is a sampling response (has result/error but no method)
+	isSamplingResponse := jsonMessage.Method == "" && jsonMessage.ID != nil &&
```
Comment (StarpTech) [inline server/streamable_http.go:240]: https://github.com/mark3labs/mcp-go/issues/554
